src/data/polynomial/mirror.lean

```diff
@@ -0,0 +1,185 @@
+- `norm2_eq_mul_reverse_coeff`: `norm2` is a coefficient of `p * p.mirror`
+
+-/
+
+namespace polynomial
+
+variables {R : Type*} [semiring R] (p : polynomial R)
+
+section mirror
+
+/-- mirror of a polynomial: reverses the coefficients while preverving `nat_degree` -/
+noncomputable def mirror := p.reverse * X ^ p.nat_trailing_degree
+
+lemma mirror_zero : (0 : polynomial R).mirror = 0 := rfl
+
+lemma mirror_monomial (n : ℕ) (a : R) : (monomial n a).mirror = (monomial n a) :=
+begin
+  by_cases ha : a = 0,
+  { rw [ha, monomial_zero_right, mirror_zero] },
+  { rw [mirror, reverse, nat_degree_monomial n a ha, nat_trailing_degree_monomial ha,
+        ←C_mul_X_pow_eq_monomial, reflect_C_mul_X_pow, rev_at_le (le_refl n),
+        nat.sub_self, pow_zero, mul_one] },
+end
+
+lemma mirror_C (a : R) : (C a).mirror = C a :=
+mirror_monomial 0 a
+
+lemma mirror_X : X.mirror = (X : polynomial R) :=
+mirror_monomial 1 (1 : R)
+
+lemma mirror_nat_degree : p.mirror.nat_degree = p.nat_degree :=
+begin
+  by_cases hp : p = 0,
+  { rw [hp, mirror_zero] },
+  by_cases hR : nontrivial R,
+  { haveI := hR,
+    rw [mirror, nat_degree_mul', reverse_nat_degree, nat_degree_X_pow,
+        nat.sub_add_cancel p.nat_trailing_degree_le_nat_degree],
+    rwa [leading_coeff_X_pow, mul_one, reverse_leading_coeff, ne, trailing_coeff_eq_zero] },
+  { haveI := not_nontrivial_iff_subsingleton.mp hR,
+    exact congr_arg nat_degree (subsingleton.elim p.mirror p) },
+end
+
+lemma mirror_nat_trailing_degree : p.mirror.nat_trailing_degree = p.nat_trailing_degree :=
+begin
+  by_cases hp : p = 0,
+  { rw [hp, mirror_zero] },
+  { rw [mirror, nat_trailing_degree_mul_X_pow ((mt reverse_eq_zero.mp) hp),
+        reverse_nat_trailing_degree, zero_add] },
+end
+
+lemma coeff_mirror (n : ℕ) :
+  p.mirror.coeff n = p.coeff (rev_at (p.nat_degree + p.nat_trailing_degree) n) :=
+begin
+  by_cases h2 : p.nat_degree < n,
+  { rw [coeff_eq_zero_of_nat_degree_lt (by rwa mirror_nat_degree)],
+    by_cases h1 : n ≤ p.nat_degree + p.nat_trailing_degree,
+    { rw [rev_at_le h1, coeff_eq_zero_of_lt_nat_trailing_degree],
+      exact (nat.sub_lt_left_iff_lt_add h1).mpr (nat.add_lt_add_right h2 _) },
+    { rw [←rev_at_fun_eq, rev_at_fun, if_neg h1, coeff_eq_zero_of_nat_degree_lt h2] } },
+  rw not_lt at h2,
+  rw [rev_at_le (h2.trans (nat.le_add_right _ _))],
+  by_cases h3 : p.nat_trailing_degree ≤ n,
+  { rw [←nat.sub_add_eq_add_sub h2, ←nat.sub_sub_assoc h2 h3, mirror, coeff_mul_X_pow',
+        if_pos h3, coeff_reverse, rev_at_le ((nat.sub_le_self _ _).trans h2)] },
+  rw not_le at h3,
+  rw coeff_eq_zero_of_nat_degree_lt (nat.lt_sub_right_iff_add_lt.mpr (nat.add_lt_add_left h3 _)),
+  exact coeff_eq_zero_of_lt_nat_trailing_degree (by rwa mirror_nat_trailing_degree),
+end
+
+lemma mirror_eval_one : p.mirror.eval 1 = p.eval 1 :=
+begin
+  simp_rw [eval_eq_finset_sum, one_pow, mul_one, mirror_nat_degree],
+  refine finset.sum_bij_ne_zero _ _ _ _ _,
```

Comment:
Is there some sort of `finset.sum_range_rev_at` that be used to golf this proof? If it doesn't exist, please leave a `TODO` comment here, but we can postpone that lemma to a future PR.

Reply:
Unfortunately not. `rev_at` is just a little thing defined in `polynomial.reverse`, and doesn't have a good API.